edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/EdgeDeploymentOperator.cs

```diff
@@ -0,0 +1,166 @@
+        readonly IEdgeDeploymentController controller;
+        readonly string iotHubHostname;
+        readonly string deviceId;
+        readonly string resourceName;
+        readonly string k8sNamespace;
+        Option<Watcher<EdgeDeploymentDefinition>> operatorWatch;
+        ModuleSet currentModules;
+
+        public EdgeDeploymentOperator(
+            string iotHubHostname,
+            string deviceId,
+            string k8sNamespace,
+            IKubernetes client,
+            IEdgeDeploymentController controller)
+        {
+            this.iotHubHostname = Preconditions.CheckNonWhiteSpace(iotHubHostname, nameof(iotHubHostname));
+            this.deviceId = Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
+            this.k8sNamespace = Preconditions.CheckNonWhiteSpace(k8sNamespace, nameof(k8sNamespace));
+            this.resourceName = KubeUtils.SanitizeK8sValue(this.iotHubHostname) + Constants.K8sNameDivider + KubeUtils.SanitizeK8sValue(this.deviceId);
+
+            this.client = Preconditions.CheckNotNull(client, nameof(client));
+            this.operatorWatch = Option.None<Watcher<EdgeDeploymentDefinition>>();
+            this.controller = Preconditions.CheckNotNull(controller, nameof(controller));
+
+            this.currentModules = ModuleSet.Empty;
+        }
+
+        public void Start()
+        {
+            // The following "List..." requests do not return until there is something to return, so if we "await" here,
+            // there is a chance that one or both of these requests will block forever - we won't start creating these pods and CRDs
+            // until we receive a deployment.
+            // Considering setting up these watches is critical to the operation of EdgeAgent, throwing an exception and letting the process crash
+            // is an acceptable fate if these tasks fail.
+            this.client.ListNamespacedCustomObjectWithHttpMessagesAsync(Constants.K8sCrdGroup, Constants.K8sApiVersion, this.k8sNamespace, Constants.K8sCrdPlural, watch: true)
+                .ContinueWith(this.OnListEdgeDeploymentsCompleted);
+        }
+
+        async Task OnListEdgeDeploymentsCompleted(Task<HttpOperationResponse<object>> task)
+        {
+            var response = await task;
+
+            this.operatorWatch = Option.Some(
+                response.Watch<EdgeDeploymentDefinition>(
+                    onEvent: async (type, item) =>
+                    {
+                        try
+                        {
+                            await this.HandleEdgeDeploymentsChangedAsync(type, item);
+                        }
+                        catch (Exception ex) when (!ex.IsFatal())
+                        {
+                            Events.ExceptionInCustomResourceWatch(ex);
+                        }
+                    },
+                    onClosed: () =>
+                    {
+                        Events.CrdWatchClosed();
```

Comment:
A really minor comment: If we've renamed everything else, we should also update the Events name...

Reply:
Good. Will do it in the next PR